src/terminal/parser/OutputStateMachineEngine.cpp

```diff
@@ -672,36 +675,52 @@ bool OutputStateMachineEngine::ActionOscDispatch(const wchar_t /*wch*/,
         break;
     }
     case OscActionCodes::SetForegroundColor:
-    {
-        std::vector<DWORD> colors;
-        success = _GetOscSetColor(string, colors);
-        if (success && colors.size() > 0)
-        {
-            success = _dispatch->SetDefaultForeground(til::at(colors, 0));
-        }
-        TermTelemetry::Instance().Log(TermTelemetry::Codes::OSCFG);
-        break;
-    }
     case OscActionCodes::SetBackgroundColor:
-    {
-        std::vector<DWORD> colors;
-        success = _GetOscSetColor(string, colors);
-        if (success && colors.size() > 0)
-        {
-            success = _dispatch->SetDefaultBackground(til::at(colors, 0));
-        }
-        TermTelemetry::Instance().Log(TermTelemetry::Codes::OSCBG);
-        break;
-    }
     case OscActionCodes::SetCursorColor:
     {
         std::vector<DWORD> colors;
         success = _GetOscSetColor(string, colors);
-        if (success && colors.size() > 0)
+        if (success)
         {
-            success = _dispatch->SetCursorColor(til::at(colors, 0));
+            size_t commandIndex = parameter;
+            size_t colorIndex = 0;
+
+            if (commandIndex == OscActionCodes::SetForegroundColor && colors.size() > colorIndex)
+            {
+                const auto color = til::at(colors, colorIndex);
+                if (color != INVALID_COLOR)
+                {
+                    success = success && _dispatch->SetDefaultForeground(color);
+                }
+                TermTelemetry::Instance().Log(TermTelemetry::Codes::OSCFG);
+                commandIndex++;
+                colorIndex++;
+            }
+
+            if (commandIndex == OscActionCodes::SetBackgroundColor && colors.size() > colorIndex)
+            {
+                const auto color = til::at(colors, colorIndex);
+                if (color != INVALID_COLOR)
+                {
+                    success = success && _dispatch->SetDefaultBackground(color);
+                }
+                TermTelemetry::Instance().Log(TermTelemetry::Codes::OSCBG);
+                commandIndex++;
+                colorIndex++;
+            }
+
+            if (commandIndex == OscActionCodes::SetCursorColor && colors.size() > colorIndex)
+            {
+                const auto color = til::at(colors, colorIndex);
+                if (color != INVALID_COLOR)
+                {
+                    success = success && _dispatch->SetCursorColor(color);
+                }
+                TermTelemetry::Instance().Log(TermTelemetry::Codes::OSCSCC);
+                commandIndex++;
+                colorIndex++;

```

Comment:
The full range is OSC 10-17. I don't think we'll be adding 13-17 in the near future. But just in case.